### ISSUE TITLE
Proposed Resolution: Change ShExJ to be compatible with JSON-LD for 2.0

### ISSUE THREAD

**Issue by gkellogg** [shexSpec/shex#18]
As discussed in #5, I propose that we change the ShExJ expression for the 2.0 release to be compatible with JSON-LD, and thus get an RDF description for ShEx.

For review, this implies a couple of relatively minor changes:

* JSON-LD does not (currently) have a way to represent the shape map (`shapeLabel` -> `shapeExpr`), and RDF certainly doesn't. This would change to be a  simple unordered array of `shapeExpr`, where the label is no the `id` (or RDF Subject) of the `shapeExpr`.
* The RDFLiteral expression in ShExJ comes straight from n3.js, and is not based on any standard. This would change to be a JSON-LD _value object_, of the form `{"@value": "literal"}` with the potential for having either `@language` or `@type` as well (all with suitable aliases we may choose).
* If the document does not contain `@context`, it is assumed to be `https://shexspec.github.io/context.jsonld`. Otherwise, using language similar to that from CSV on the Web, it must be that, or an array who's first entry is that IRI with the other entry a JSON Object containing `@base` and simple term definitions that take the place of `base` and `prefixes`.
* `prefixes` and `base` in ShExJ is unused in reality and appear mostly for documentation purposes. We could retain them as part of the `@context`.

Additionally, the [JSON-LD Context](https://shexspec.github.io/context.jsonld) includes an RDFS definition for the ShEx Namespace: `https://shexspec.github.io/ns/`, which you can also find in Turtle at https://shexspec.github.io/ns/shex.ttl.

You can express your support (or otherwise) of this proposal either with a positive comment, or using the GitHub reaction interface above.

**Comment by labra** [comment]
👍

**Comment by jimkont** [comment]
Hi, I am a new-comer here, apologies if I miss something obvious or something that is already discussed. What is the motivation behind this issue? to be able to just-store shex in a triple-store or have an RDF representation for ShEx?

If it is the latter, IMO there should be some definitions of the existing RDF syntax and examples of how all the existing and 2.0+ features are serialised in RDF (the links seem outdated).
Once the desired RDF syntax is complete, the JSON-LD could be easy & safe to create.

Currently, it looks like this is going the opposite way and there could be some edge cases where a mapping would not be possible. 
On the other hand, we have json-ld and shex experts that agree here :)

**Comment by labra** [comment]
I will try to explain my vote and on how things evolved: 

- ShEx was originally defined using an [abstract syntax](https://www.w3.org/2013/ShEx/Definition) which enabled it to have several serialization formats like ShExC (compact notation), RDF or JSON.
- There was some interest in a ShEx-RDF serialization format and there is some historical discussion about it [here](https://www.w3.org/2001/sw/wiki/ShEx/RDF_serialization). However, there was not enough interest or resources to maintain it and it got a bit abandoned.
- For version 2.0 of ShEx, we considered that the abstract syntax could be defined in terms of Json, and Eric developed [ShExJ](https://shexspec.github.io/spec/#shape-expressions-shexj) which could serve as both the abstract syntax and a Json based syntax. 
- In that way, some ShEx processors could implement ShEx without the need to implement a parser for the compact syntax. As far as I know, that's what Iovka's processor does. 
- Given that having an RDF serialization format was seen as a [problem for adopting ShEx](https://lists.w3.org/Archives/Public/public-rdf-shapes/2016Dec/0061.html), I added issue #5 to resume our previous work on RDF serialization format. 
- Gregg proposed a very interesting approach which was to change ShExJ to be JSON-LD compatible. In that way, we could kill 2 (or 3) birds with one stone: having an abstract syntax for ShEx, with a JSON and RDF serialization format for free. 

The needed changes for ShExJ seem plausible, and that's why I voted +1. because I think this approach is both elegant and useful. 

In a nutshell, you are right that some links may be outdated mainly because we still need to pollish the definitions. As I can see it, the definitions would be in the [spec](https://shexspec.github.io/spec/#shape-expressions-shexj) document once it gets updated to reflect the changes proposed by Gregg.

**Comment by gkellogg** [comment]
@jimkont I think there are a couple of motivations:

1) ShExJ was not really based on any existing ShEx format, and as it happens, it is _almost_ JSON-LD.
2) There has been a desire to have an RDF expression of ShEx, and given that JSON-LD is an RDF syntax, this seemed like a natural. RDF provides a much better basis for a data model, as JSON has no understanding of things like RDF Literals or IRIs.
3) By using a constrained subset of JSON-LD, we can continue to allow for JSON-specific processors to be able to use the shape of the JSON predictably. And, with JSON-LD framing, we can actually take and arbitrary RDF syntax, re-serialize it to JSON-LD and perform JSON-LD framing to get back precisely the shape of the JSON we need. Although, in theory, one could do things using RDF queries on a graph, the fact that you can predictably go from an abstract RDF graph to framed JSON-LD means that the work is made much easier.
4) JSON-LD was specifically designed to "bridge the gap" between RDF and an easy to use JSON serialization, so we gain the strong theoretical basis of RDF and RDFS while presenting developers with an easy to use serialization.

> IMO there should be some definitions of the existing RDF syntax and examples of how all the existing and 2.0+ features are serialised in RDF (the links seem outdated).

All examples in the ShEx spec have been translated to JSON-LD, which you can see here: https://github.com/shexSpec/spec/tree/gh-pages/examples. In fact, I think @ericprud has automatically changed all of the test-suite files as well; if not, it should be pretty trivial.

> Currently, it looks like this is going the opposite way and there could be some edge cases where a mapping would not be possible.

In what way do you you see this as going in the opposite way? I believe all existing examples are handled with only minor variation.

In fact, one thing that could be simplified is Inclusion, where the reference is to a Shape, from which you need to find a TripleConstraint. With RDF/JSON-LD we could simply reference the TripleConstraint directly using it's `id`. This is presumably the purpose of `productionLabel`, which is otherwise unused in the grammar. In fact, all labeling is simplified by the fact that all nodes in RDF have a subject, be it an IRI or a BNode.

**Comment by jimkont** [comment]
Thank you both for your answers, I was obviously missing the provided background info.

**Comment by gkellogg** [comment]
This is complete now.
